### PR TITLE
Add velocity support

### PR DESCRIPTION
**Pull request by MrIvanPlays**
This rewrites #39 as it looks like the author of it does not care about it anymore. Ignore the commit message, its because i type very fast and idk why my keyboard didnt handle the "y".

### CODE REVIEW

**Comment by MrIvanPlays** [comment]
3 months and no response??

**Comment by Bastian** [comment]
Hi,
sorry for the slow response, but like I've already said in the other PR,I just don't have the time to properly test any bigger PRs currently. Especially a new Metrics class needs an in-depth review and testing, as any further changes to this files must be 100% backwards-compatible.

File: bstats-velocity-lite/src/main/java/org/bstats/velocity/MetricsLite.java [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+     */
+    public JsonObject getPluginData() {
+        return pluginData.getSent();
+    }
+
+    private void startSubmitting() {
+        // The data collection is async, as well as sending the data
+        // Velocity does not have a main thread, everything is async
+        pluginData.getServer().getScheduler().buildTask(pluginData.getPluginInstance(), this::submitData)
+                .delay(2, TimeUnit.MINUTES).repeat(30, TimeUnit.MILLISECONDS).schedule();
```
Comment (Bastian) [inline bstats-velocity-lite/src/main/java/org/bstats/velocity/MetricsLite.java:152]: This code sends the data to bStats every 30 milliseconds.

**Comment by Bastian** [comment]
Whoever is already using this Metrics class in their plugin: Stop it! As long as a Metric class is not approved by me, you are not allowed to use it to send data to bStats.
I don't want to see stuff like this in my logs:
image

**Comment by Andre601** [comment]
Would there be any specific requirements/procedures to test this? I would gladly offer my help in testing this out to make sure it meets the requirements (And to get it closer to a possible merge)

And while I'm at it, do I mention here, what I did in #81 which is to perhaps consider trying to find a platform-independant way of creating a metrics class, where instead of providing something like a Plugin instance, you could provide more generic info which allows the usage of a central metrics class, which would adabt itself to the platform that the plugin is currently used on. This could essentially reduce and/or eliminate multiple copy-paste code of pretty much the same stuff.

**Comment by Bastian** [comment]
Maybe a small update about my future plan for the Metrics classes in general: At the moment, the most challenging part about reviewing and updating Metrics classes is the fact, that all plugins have to communicate between each other to send a single request to the bStats server (instead of one request per plugin).

This comes with some major drawbacks because:
1. Every version needs to be forwards **and** backwards-compatible. If there's a server with a plugin that uses a 4 year old Metrics class and a recent one, they must be able to communicate. 
2. Communicating between the other plugins is complicated and depends on the platform:
    * For Bukkit, bStats (ab)uses the built-in [ServicesManager](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/plugin/ServicesManager.html).
    * For Bungeecord, bStats uses a temporary text file that contains the fully-qualified class name of the first-loaded Metrics class. 
    * For Spone, it's a real mess: It initially used a temporary text file like Bungeecord and then switched to using Sponge's service manager (I don't even know how it works. The Metrics class was re-written by a Sponge developer). To make it compatible with the older version, the Sponge Metrics class in now super complicated ([see here)(https://github.com/Bastian/bStats-Metrics/blob/master/bstats-sponge-lite/src/main/java/org/bstats/sponge/MetricsLite2.java#L278-L332)).
    * The Metrics class in the PR appears to use a temporary text file like Bungeecord.
3. Changes to the platform API itself can cause problems. E.g., when Spigot removed the `json-simple` dependency, the new Metrics class had to use some ugly reflection to maintain compatability ([See here](https://github.com/Bastian/bStats-Metrics/blob/4a781501b6c839f98978d82c91eeed9f32e7a72b/bstats-bukkit-lite/src/main/java/org/bstats/bukkit/MetricsLite.java#L267-L287)).

The main reason for grouping up multiple requests into a single one is performance. However, I'm currently trying to evaluate the performance overhead of one request per plugin. I don't think it will be a very significant performance impact but it would greatly simplify the Metrics classes.

If it turns out, that one request per plugin is feasibly, I plan on making a common codebase for all platform-independent stuff (e.g., data sending, custom charts, ...) and a smaller platform-dependent class that only deals with data-collection and features like Sponge's built-in Metrics opt-in meachanism. Such a Metrics class would allow me to add support for new platforms (like Velocity) very easily.

**Comment by Bastian** [comment]
I also hope that this change would allow for an easier expansion to complete new areas like pocket-edition servers or client mods.

**Comment by Andre601** [comment]
Perhaps communicating with the different platforms would be an approach to find a solution?

It could be beneficial if the platforms could offer a api for plugins to add their own bstats statistics.
That way could plugins provide values on their enable and the server/proxy could collect and send those at once when it is ready.

**Comment by Bastian** [comment]
I don't think that's a good idea.
1. This would make the Metrics classes more platform-dependant again -> That's exactly what I want to move away from.
2. Not every platform will support this. E.g., if you know md_5, you should know that there's now way that he would add something like this to Spigot.
3. What's about the existing Metrics classes and older Minecraft versions?

**Comment by Andre601** [comment]
Respect old versions all you want, there is a moment where you have to make breaking changes.
That's one reason why RestAPIs often have dedicated versions for api paths (example.com/api/v1/...).

So you should perhaps consider to start making metrics stuff version specific...

**Comment by Andre601** [comment]
An additional alternative would perhaps be to just offer an official BStats-plugin for the server platforms, which the plugins could then hook into.

Possible Benefits:
- Central API to use for plugins to submit data.
- Plugins could keep their Metrics class as backup and have a simple check for if the bstats-plugin is available, to otherwise use their metrics class.
- Plugins keeping their metrics class and not supporting the plugin *should* still work independantly.

Downsides:
- Would be yet another plugin people need to install.
- Because of above would it essentially switch from an opt-out to an opt-in system (Possible heavy loss of stats).
- 4+ different plugins would need to be made. One for each server type, unless platform independant ways can be found to retrieve the data.

**Comment by Bastian** [comment]
But what's wrong with my suggested switch to one-request-per-plugin? It would completely eliminate the need for the Metrics class to be backwards compatible.

A few hours ago, I've deployed an update that splits the current incoming requests into smaller one-per-plugin batches and I'm seeing an ~50% increase in CPU usage. While +50% is far from perfect (and maybe can still be optimized a little bit), it's still reasonable and probably worth the benefits.

> Even further do I not see a possible way to keep the submitting of the data secure in a way. Like any plugin that has the plugin ID could submit any data they want and therefore destroy the stats of that legit plugin, so there is definitely space for some improvements.

This is always possible. There's no way for bStats to know whether a request is coming from a legitimate source or not. The only possibility is to filter out suspicious requests, e.g. when a server has 10000 players.

**Comment by Andre601** [comment]
> But what's wrong with my suggested switch to one-request-per-plugin? It would completely eliminate the need for the Metrics class to be backwards compatible.

The biggest issue I would see with that is possible rate limits from a possible host, but not sure.
Some hosts *love* to limit requests towards the internet... I think.
It's a minor problem tho.

**Comment by Andre601** [comment]
I for my part would really suggest to perhaps really consider to go with a plugin, as through that would it stay 1 request, it would always be the same plugin collecting and sending the info (Therefore won't have a large chance of messing up the JSON or whatever) and info like server type, version, etc could be collected by the plugin while the others would only provide their own information (Going more towards a platform-independent system, as the plugin would collect server specific info.)

But at the end would it be up to you to decide.

**Comment by MrIvanPlays** [comment]
a plugin would make it even more platform dependent, what bastian wants to move from too.

**Comment by Bastian** [comment]
Hmm, this was automatically closed by force-pushing to master and I cannot re-open it.

**Pull request closed without merging**